src/EventProducerService.cpp

```diff
@@ -0,0 +1,217 @@
+// Copyright (C) Martin Da Costa 2023 (martindc.merg@gmail.com)
+// This file is part of VLCB-Arduino project on https://github.com/SvenRosvall/VLCB-Arduino
+// Licensed under the Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License.
+// The full licence can be found at: http://creativecommons.org/licenses/by-nc-sa/4.0/
+
+// TODO: Add AREQ and ASRQ opcode support.
+// Check error messages
+// Set Params flags
+
+#include <Streaming.h>
+#include "EventProducerService.h"
+#include <Controller.h>
+#include <vlcbdefs.hpp>
+
+namespace VLCB {
+
+void EventProducerService::setController(Controller *cntrl) 
+{
+  this->controller = cntrl;
+  this->module_config = cntrl->module_config;
+}
+
+void EventProducerService::begin()
+{
+  //Initialise instantMode
+  instantMode = module_config->currentMode;
+  //DEBUG_SERIAL << F("> instant MODE initialise as: ") << instantMode << endl;
+  
+  if (instantMode == MODE_UNINITIALISED)
+  {
+    return;
+  }
+  else
+  {
+    setProducedEvents();
+  }
+}
+
+void EventProducerService::setProducedEvents()
+{
+  if (module_config->EE_PROD_EVENTS > 0)
+  {
+    for (byte i = 0; i < module_config->EE_PROD_EVENTS; i++)
+    {
+      if (module_config->getEvTableEntry(i) == 0)
+      {
+        // Event does not exist so create it
+        byte data[4];
+        data[0] = highByte(module_config->nodeNum);
+        data[1] = lowByte(module_config->nodeNum);
+        data[2] = 0;
+        data[3] = i;
+        
+        module_config->writeEvent(i, data);  //index = i
+        module_config->writeEventEV(i, 1, 1);
+        module_config->writeEventEV(i, 2, 0);  //set default switch type
+        module_config->updateEvHashEntry(i);
+      }
+    }    
+  }
+}
+
+void EventProducerService::process(byte num)
+{
+  // Do this if mode changes
+  instantMode = controller->newMode;
+  if (lastMode != instantMode)
+  {
+    // Do this if change from setup to normal.
+    if ((lastMode == MODE_SETUP) && (instantMode == MODE_NORMAL))
+    {
+      setProducedEvents();
```

Comment:
Isn't it enough to call setProducedEvents() once, i.e. from begin().

Reply:
No, because if there has been a reset to manufacturers defaults (OPC_NNRSM) the vents table will have been cleared and the default produced evets will need to be re-written.

Reply:
Interesting edge case. So if I understand you there is no need to do this every time we leave setup mode. Possibly there should be another mechanism where things are set up after a reset (or from Uninitialized.)

Reply:
Actually, we don't need it!  After a resetModule() we reboot()! (Configuration.cpp line 567 - 597).  I'll take it out.

Reply:
That leaves process() empty so I'll lose that, as well.